statistics/table.go

```diff
@@ -155,74 +170,87 @@ func (t *Table) String() string {
 	return strings.Join(strs, "\n")
 }
 
-// ColumnIsInvalid checks if this column is invalid.
-func (t *Table) ColumnIsInvalid(colInfo *model.ColumnInfo) bool {
+type neededColumnMap struct {
+	m    sync.Mutex
+	cols map[int64][]*Column
+}
+
+var neededColumns = neededColumnMap{cols: map[int64][]*Column{}}
+
+// ColumnIsInvalid checks if this column is invalid. If this column has histogram but not loaded yet, then we mark it
+// as need histogram.
+func (t *Table) ColumnIsInvalid(colID int64) bool {
 	if t.Pseudo {
 		return true
 	}
-	col, ok := t.Columns[colInfo.ID]
+	col, ok := t.Columns[colID]
+	if ok && col.NDV > 0 && len(col.Buckets) == 0 {
+		neededColumns.m.Lock()
+		neededColumns.cols[t.TableID] = append(neededColumns.cols[t.TableID], col)
```

Comment:
It's possible that the same column being appended multiple times before the histogram is loaded.
We can use a struct 
```
type TableColumnID {
    TableID int64
    ColumnID int64
}
```
As the key of the map `map[TableColumnID]struct{}`